src/extension.ts

```diff
@@ -48,45 +48,71 @@ async function fetchPackageInfo(requirement: PackageRequirement): Promise<Packag
+    const infoPresentation: Array<string> = [
+        headPart, authorSubpart + licenseSubpart, versionPart, versionPartLens
+    ].filter(Boolean)
+
     return infoPresentation
 }
 
 async function provideHover(document: vscode.TextDocument, position: vscode.Position) {
     const requirement: PackageRequirement | null = extractPackageRequirement(document.lineAt(position.line))
     if (requirement === null) return new vscode.Hover('')
-    let infoPresentation: string | undefined = infoPresentationCache.get(requirement.id)
-    if (infoPresentation === undefined) {
-        const [status, info]: PackageInfoRequest = await fetchPackageInfo(requirement)
-        if (status === 200) infoPresentation = presentPackageInfo(info!)
-        else if (status === 404) infoPresentation = `{id_raw} is not available in PyPI`
-        if (infoPresentation) infoPresentationCache.set(requirement.id, infoPresentation)
-        else infoPresentation = linkify(
-            `could not fetch ${requirement.id_raw} information from PyPI`,
-            `https://pypi.org/project/${requirement.id_raw}/`
-        )
-    }
-    infoPresentation = infoPresentation.replace('{id_raw}', requirement.id_raw)
-    return new vscode.Hover(infoPresentation)
+    let infoPresentation: Array<string> | undefined = infoPresentationCache.get(requirement.id)
+    if (infoPresentation === undefined) return new vscode.Hover('')
+    return new vscode.Hover(infoPresentation.slice(0,-1).join("\n\n").replace('{id_raw}', requirement.id_raw))
 }
 
+class MyCodeLensProvider implements vscode.CodeLensProvider {
+    async provideCodeLenses(document: vscode.TextDocument): Promise<vscode.CodeLens[]> {
+
+      const lineCount: number = document.lineCount;
+
+      const requirements: Array<PackageRequirement|null> = 
+      [...Array(lineCount).keys()]
+      .map(lineNumber => extractPackageRequirement(document.lineAt(lineNumber)))
+      
+      const infoPresentations: Promise<vscode.Command[]> = Promise.all(requirements.map(async (requirement)=>{
+        if(requirement === null) return { command: "", title: ""}
+        let infoPresentation: Array<string> | undefined = infoPresentationCache.get(requirement.id)
+        if (infoPresentation === undefined) {
+          const [status, info]: PackageInfoRequest = await fetchPackageInfo(requirement)
+          if (status === 200) infoPresentation = presentPackageInfo(info!)
+          else if (status === 404) infoPresentation = [`{id_raw} is not available in PyPI`]
+
+          if (infoPresentation) infoPresentationCache.set(requirement.id, infoPresentation)
+          else infoPresentation = [linkify(
+              `could not fetch ${requirement.id_raw} information from PyPI`,
+              `https://pypi.org/project/${requirement.id_raw}/`
+          )]
+        }
+        return {command: "", title: infoPresentation.slice(-1)[0].replace('{id_raw}', requirement.id_raw)}
+      }))
```

Comment:
This seems more complicated than it really should because error handling was moved from the hover provider to here. However, I'd rather we just didn't show anything if there was an error, in order to reduce visual clutter (a ton of "could not fetch" messages all at once would disrupt the workflow more than it would help).

Reply:
I've hidden the errors and placed the old error messages as comments.